rp2040-hal/src/timer.rs

```diff
@@ -106,6 +112,75 @@ impl Timer {
 // safety: all write operations are synchronised and all reads are atomic
 unsafe impl Sync for Timer {}
 
+/// [`Timer`] based Delay
+pub struct Delay(());
+impl Delay {
+    fn delay_from_u32(&self, us: u32) {
+        // safety: only reads are made
+        let timer = unsafe { &*pac::TIMER::PTR };
+        let start = timer.timelr.read().bits();
+        while timer.timelr.read().bits().wrapping_sub(start) <= us {
+            // use nop instead of wfe because there is no event configured to wake us up.
+            cortex_m::asm::nop();
+        }
+    }
+}
+macro_rules! impl_delay {
+    (u32: $($t:ty),*) => {
+        $(impl embedded_hal::blocking::delay::DelayUs<$t> for Delay {
+            fn delay_us(&mut self, us: $t) {
+                self.delay_from_u32(u32::from(us))
+            }
+        })*
+        $(impl embedded_hal::blocking::delay::DelayMs<$t> for Delay {
+            fn delay_ms(&mut self, ms: $t) {
+                for _ in 0..ms {
+                    self.delay_from_u32(1_000)
+                }
+            }
+        })*
+    };
+}
+
+impl_delay!(u32: u8, u16, u32);
+impl embedded_hal::blocking::delay::DelayUs<u64> for Delay {
+    fn delay_us(&mut self, us: u64) {
+        use embedded_hal::blocking::delay::DelayMs;
+
+        let (ms, us) = (us / 1_000, us % 1000);
```

Comment:
According to figure 2 on page 26 of https://datasheets.raspberrypi.com/pico/raspberry-pi-pico-c-sdk.pdf, a 64 bit division using the hardware divider takes about 100 cycles. (Assuming that's the unit used in the graph.)
That's 0.8µs at 125MHz. Even if the compiler is smart enough to merge the division and the modulo operation into a single division, the CPU actually runs at 125MHz, and the hardware divider is used, this means that a 1µs delay becomes significantly longer than needed.
Given that the timer has a 64bit counter, why do you do this division at all? Is it just to avoid overflow issues in the calculations?

Reply:
Good points! I actually completely missed that his implementation is only used for u64. Should be more careful during code review.
While it's theoretically possible to use the u64 version to do a 1µs delay, performance-sensitive callers would probably use a more suitable type and won't be affected by the division.

Reply:
I had a look at the generated assembly. While it doesn't matter at all whether the parameter of the inner function is an u32 or an u64 (the call gets inlined anyway), doing the subtraction and comparison math in u32 saves several instructions.